### ISSUE TITLE
Upgrade from rails 6.1 to 7.0 - cookie rotator doesn't make it possible to read old cookies.

### ISSUE THREAD

**Issue by bubiche** [rails/rails#52288]
### Steps to reproduce

- Upgrade rails gem version from 6.1.7.8 to 7.0.8.4
- Add the cookie rotator as described in https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#key-generator-digest-class-change-requires-a-cookie-rotator
```ruby
# config/initializers/cookie_rotator.rb
Rails.application.config.after_initialize do
  Rails.application.config.action_dispatch.cookies_rotations.tap do |cookies|
    authenticated_encrypted_cookie_salt = Rails.application.config.action_dispatch.authenticated_encrypted_cookie_salt
    signed_cookie_salt = Rails.application.config.action_dispatch.signed_cookie_salt

    secret_key_base = Rails.application.secret_key_base

    key_generator = ActiveSupport::KeyGenerator.new(
      secret_key_base, iterations: 1000, hash_digest_class: OpenSSL::Digest::SHA1
    )
    key_len = ActiveSupport::MessageEncryptor.key_len

    old_encrypted_secret = key_generator.generate_key(authenticated_encrypted_cookie_salt, key_len)
    old_signed_secret = key_generator.generate_key(signed_cookie_salt)

    cookies.rotate :encrypted, old_encrypted_secret
    cookies.rotate :signed, old_signed_secret
  end
end
```

### Expected behavior

- Users who already logged in before the upgrade stays logged in.

### Actual behavior

- Users are logged out.
- The issue happens in my local environment so I hope to be able to make sure it works when deploying rails 7 to users.

### System configuration
**Rails version**: 7.0.8.4

**Ruby version**: 3.2.2

**Comment by Earlopain** [comment]
I've used this exact same rotator a few years ago and that worked fine. Can you make sure that `Rails.application.config.active_support.hash_digest_class` before your upgrade is set to `OpenSSL::Digest::SHA1`? That's the only thing I can think of that can could make trouble here. 1000 iterations as well, though I'm not sure how you'd even configure that. Similarly, for development `tmp/development_secret.txt`/`Rails.application.secret_key_base` must be the same.

**Comment by bubiche** [comment]
Thanks for the response! Both `Rails.application.config.active_support.hash_digest_class` and `Rails.application.config.active_support.key_generator_hash_digest_class` are `nil` for me instead of `OpenSSL::Digest::SHA1` before the upgrade. Would this be what caused my issue?

I don't do any extra config for cookies so I think iterations should be the same, `secret_key_base` are the same too.

**Comment by Earlopain** [comment]
Nah, this is my bad. It makes sense that these return nil because rails doesn't know about these configs yet in that version.

Do you already load 7.0 defaults/uncommented  the relevant configs in the upgrade file? I'm not sure how this would behave if it is still set to `SHA1` while using the cookie rotator.

If you don't do anything special for cookies I'm not really sure what else you could look at :shrug:

**Comment by bubiche** [comment]
I uncommented the relevant configs in the upgrade file. I assume they are `active_support.key_generator_hash_digest_class` and `active_support.hash_digest_class` right?

**Comment by bubiche** [comment]
For those who somehow got the same issue, I have found the reason, for some reason I have the config `use_authenticated_cookie_encryption` set to `false` in my development environment so my cookies encryptor/cipher is different:

https://github.com/rails/rails/blob/b84fa1083ca7e6422356fdf80cb33751d0a23eff/actionpack/lib/action_dispatch/middleware/cookies.rb#L659

I have to do this instead: `cookies.rotate :encrypted, old_encrypted_secret, old_signed_secret`

**Comment by clucas** [comment]
I suspect that you upgraded your app from Rails 5.2 without enabling this setting in the new_framework_defaults_5_2.rb
`Rails.application.config.action_dispatch.use_authenticated_cookie_encryption = true`
